ocaml/xapi/message_forwarding.ml

```diff
@@ -3982,13 +3982,10 @@ functor
             Client.Host.set_https_only ~rpc ~session_id ~self ~value
         )
 
-      let apply_recommended_guidances ~__context ~self =
-        let uuid = host_uuid ~__context self in
-        info "Host.apply_recommended_guidance: host = %s" uuid ;
-        let local_fn = Local.Host.apply_recommended_guidances ~self in
-        do_op_on ~local_fn ~__context ~host:self (fun session_id rpc ->
-            Client.Host.apply_recommended_guidances ~rpc ~session_id ~self
-        )
+      let apply_recommended_guidances ~__context ~host =
+        let uuid = host_uuid ~__context host in
+        info "Host.apply_recommended_guidances: host = %s" uuid ;
+        Local.Host.apply_recommended_guidances ~__context ~host
     end
```

Comment:
Why is this correct if host not equal to master? This needs an comment why we don't need to forward this to another host.

Reply:
This is the purpose of this PR: it is because we "can not" forward this to pool member, as the pool member is running with an old version of XAPI which does not have the new added API call at that time. I am afraid it is had to make it clear with an comment, below is the detailed scenario(as [mentioned in the design doc](https://info.citrite.net/pages/viewpage.action?spaceKey=xenserver&title=Design+of+Toolstack+for+CP-41451#comment-1631270801)):

let's discuss pool upgrading with a pool with 2 hosts: host A as master, B as slave, both with old xapi version (without the new host.apply_recommended_guidances  API):


step 1: XenCenter call host.apply_updates  on master A, as master XAPI is old, after it installed the new version of xapi-core rpm, it restart xapi immediately as recommended guidance of the update. As toolstack is restarted, XenCenter will not get the result of the host.apply_updates . So it will not got the Api_errors.updates_require_recommended_guidance  warning, thus user will not get the recommended guidance prompt.


step 2: After XenCenter reconnect with the pool, it continues to call host.apply_updates  on host B, here is the place which I missed: the implementation of apply_updates  handler is that, the apply_updates  request is not proxied from master A to host B directly, now most of the handling of this request is executed in master, since now master A has been upgraded to new version, it will not apply recommended guidance within host.apply_updates , and the new version XAPI on master will also respond warning: Api_errors.updates_require_recommended_guidance  for the host.apply_updates  API call, then client would try to call the new API: host.apply_recommended_guidances . The currently implementation of host.apply_recommended_guidances  handler is that: master will simply proxy this request to slave, while XAPI running in slave is still old as toolstack on slave have not been restarted yet, then the host.apply_recommended_guidances  request would fail.